### ISSUE TITLE
Performance issues training SlotFormer

### ISSUE THREAD

**Issue by alexcbb** [pairlab/SlotFormer#11]
Hello,

I've been trying to launch the training code for SlotFormer with SAVi on CLEVRER on DDP with 2GPUS (RTX 6000 of 24GB) and I found out that the training time was really long (~12h per epoch). As the training is 80 epochs, it would be a total training time of 40 days... (VS the 5 days announced). I've already tried to change the num_workers to different values, but it does not seems to have a big impact over the training time. 

Can you provide me the exact setup you trained the 80 epochs models (num gpus, batch_size, total time, etc.) ?

Thank you in advance.

**Comment by Wuziyi616** [comment]
I don't think there is anything special in our cluster setup -- 2 RTX 6000 GPUs with 24 GB each, and PyTorch >= 1.10. The batch_size etc. are the same as in the config file. What's your runtime for each training step? Can you check the GPU utils and see if it's high (e.g. >90%)?

**Comment by alexcbb** [comment]
For 1 training step with 2 gpus, the forward varies from 0.5s to 5s and data loading varies from some ms to 6s. With 1 GPU the forward and data loading is very stable (around 0.3s) and the data loading is around 5s. I believe there might be some issues with GPU synchronization or communication but still, the setup with one GPU is not that fast either.

**Comment by Wuziyi616** [comment]
> data loading varies from some ms to 6s

I don't know why forward time can vary, but at least this data loading time sounds like a big issue to me. Have you preprocess videos to save as individual frames? It will greatly speed up data loading. See [these lines](https://github.com/pairlab/SlotFormer/blob/master/slotformer/base_slots/datasets/clevrer.py#L73-L92). Say in the original dataset you have a video called `clevrer/train/0000.mp4`, then you can save all of its frames to a folder called `clevrer/train/0000/`.

**Comment by Wuziyi616** [comment]
I've attached a script I used for preprocessing videos.
[clevrer_video2frames.txt](https://github.com/user-attachments/files/16106731/clevrer_video2frames.txt)

**Comment by alexcbb** [comment]
Ok I see, you're right I didn't thought about this but it may be a solution. Did you made some evaluation on how much it speeds up the loading ? I was also thinking about converting the dataset to a format such as h5py in order to even increase the training speed. Thank you again for your help. I'll see with this if I still got the same issue.

**Comment by Wuziyi616** [comment]
I don't remember the exact speed up but data loading should be at the ms level I believe. Feel free to reopen the issue if this doesn't help.

**Comment by alexcbb** [comment]
Hello, related to this I have found something weird in the training forward with nerv. For the BaseMethod, in the training_epoch, there is a definition of train step which is the following :
```
train_steps = (
            len(self.train_loader.sampler) -
            self.train_loader.sampler.real_counter(self.iter_train_loader)
        ) // (self.params.train_batch_size * self.gpus)
```

For the dataset CLEVRER there is 721079 instances. I launched a DDP training with a batch size of 64 splitted over 2 gpus (so 32 per gpu) and this train_steps value is showing 5633 per GPU (with the following tqdm) which seems wrong to me. 721079 / 64 == 11 266. How can it be at 5633 ? As the sampler is already splitting the data accross the gpus, there is no need to multiply by "self.gpus" I guess. I indeed tried to launch the training without nerv but with a classical Pytorch loop and classical DistributedSampler and I'm not able to have a dataloading as fast as I have with your repo. Do you have maybe any clue and if this could be related ?

**Comment by Wuziyi616** [comment]
There was an error in the calculation of `train_steps` in the previous version of `nerv` (specifically, v0.1.0 which is specified in the README). I wasn't dividing by `self.gpus` previously so when using multiple GPUs it will be problematic. This is fixed in [this commit](https://github.com/Wuziyi616/nerv/commit/ccc8a8ddb21a2561649c4651818497b8ed60a9bc). Not sure what's the version you installed (maybe check with `pip show nerv`). Overall I don't think this will affect the model training as long as you are training for enough steps.

**Comment by Wuziyi616** [comment]
> As the sampler is already splitting the data accross the gpus, there is no need to multiply by "self.gpus" I guess

Oh I didn't think about this. Well I guess it won't affect the model performance as long as you train for enough steps? Because the data batches are shuffled anyways.

Regarding `* self.gpus`, it is weird cause IIRC I added it when I saw wrong step count in one of my previous experiments. I recall the step shows 10k steps, but the training only runs for 5k steps and then raise a `StopIterationError`. Very curious...

**Comment by alexcbb** [comment]
Yes, I think you are right, but what bothers me is the fact that between 2 epochs you might have seen different data as the model don't go over the entire dataset. I've removed the `*self.gpus` on my current version, now I've got the right number of steps (~11k steps) and the time do not seems to have changed for the full forward pass over the dataiterator. I did a try on my side and did not met the `StopIterationError`. But now, I'm also still wondering why I'm not able to reproduce the same behavior with pure Pytorch (the next() of the data iterator is quite fast with nerv, ~40min per epoch, but not with an original Pytorch loop, ~3h per epoch, I'm expecting either an error in my reproduction or something weird with the nerv loop)

**Comment by alexcbb** [comment]
Shall I make a PR on nerv repo for this issue by the way ?

**Comment by Wuziyi616** [comment]
> (the next() of the data iterator is quite fast with nerv, ~40min per epoch, but not with an original Pytorch loop, ~3h per epoch

Very curious indeed. I don't think I have anything fancy on the data side. Mind sharing your pure PyTorch code? Also, have you tried the single-GPU (no DDP) case, does that give you the same speed?

**Comment by Wuziyi616** [comment]
Would be great if you can make a PR for the `* gpus` thing and link to this issue. Thanks!

**Comment by alexcbb** [comment]
> > (the next() of the data iterator is quite fast with nerv, ~40min per epoch, but not with an original Pytorch loop, ~3h per epoch
> 
> Very curious indeed. I don't think I have anything fancy on the data side. Mind sharing your pure PyTorch code? Also, have you tried the single-GPU (no DDP) case, does that give you the same speed?

For the dataset I use exactly the same code, but instead of using the BaseDataModule I use a classical DistributedSampler and Pytorch Dataloader (with the same parameters and num of workers that the original script). The thing is, the last time I tried for single GPU it was also quite slow. But nevermind, as everything seems to work with the original repo I will continue to work on top of it for my own experiments.
